django/db/models/fields/tuple_lookups.py

```diff
@@ -0,0 +1,244 @@
+import itertools
+
+from django.core.exceptions import EmptyResultSet
+from django.db.models.expressions import ColPairs, Func, Value
+from django.db.models.lookups import (
+    Exact,
+    GreaterThan,
+    GreaterThanOrEqual,
+    In,
+    IsNull,
+    LessThan,
+    LessThanOrEqual,
+)
+from django.db.models.sql.where import AND, OR, WhereNode
+
+
+class Tuple(Func):
+    function = ""
+
+
+class TupleLookupMixin:
+    def get_prep_lookup(self):
+        self.check_tuple_lookup()
+        return super().get_prep_lookup()
+
+    def check_tuple_lookup(self):
+        assert isinstance(self.lhs, ColPairs)
+        self.check_rhs_is_tuple_or_list()
+        self.check_rhs_length_equals_lhs_length()
+
+    def check_rhs_is_tuple_or_list(self):
+        if not isinstance(self.rhs, (tuple, list)):
+            raise ValueError(
+                f"'{self.lookup_name}' lookup of '{self.lhs.field.name}' field "
+                "must be a tuple or a list"
+            )
+
+    def check_rhs_length_equals_lhs_length(self):
+        if len(self.lhs) != len(self.rhs):
+            raise ValueError(
+                f"'{self.lookup_name}' lookup of '{self.lhs.field.name}' field "
+                f"must have {len(self.lhs)} elements"
+            )
+
+    def check_rhs_is_collection_of_tuples_or_lists(self):
+        if not all(isinstance(vals, (tuple, list)) for vals in self.rhs):
+            raise ValueError(
+                f"'{self.lookup_name}' lookup of '{self.lhs.field.name}' field "
+                f"must be a collection of tuples or lists"
+            )
+
+    def check_rhs_elements_length_equals_lhs_length(self):
+        if not all(len(self.lhs) == len(vals) for vals in self.rhs):
+            raise ValueError(
+                f"'{self.lookup_name}' lookup of '{self.lhs.field.name}' field "
+                f"must have {len(self.lhs)} elements each"
+            )
```

Comment:
Why we don't have any tests for these checks? Secondly, some of them are used only in a single lookup so I'm not sure why there are in the mixin :thinking:, and, as far as I'm aware, some of them are unnecessary, e.g. `check_rhs_is_tuple_or_list()`.

We also should not call functions (e.g. `len()`) in f-strings as described in the contributing guide.